core/src/main/java/io/confluent/kafka/schemaregistry/storage/InMemoryCache.java

```diff
@@ -127,4 +129,92 @@ public void schemaRegistered(SchemaKey schemaKey, SchemaValue schemaValue) {
   public List<SchemaKey> deletedSchemaKeys(SchemaValue schemaValue) {
     return guidToDeletedSchemaKeys.getOrDefault(schemaValue.getId(), Collections.emptyList());
   }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public AvroCompatibilityLevel compatibilityLevel(String subject,
+                                                   boolean returnTopLevelIfNotFound,
+                                                   AvroCompatibilityLevel defaultForTopLevel
+  ) {
+    ConfigKey subjectConfigKey = new ConfigKey(subject);
+    ConfigValue config = (ConfigValue) get((K) subjectConfigKey);
+    if (config == null && subject == null) {
+      return defaultForTopLevel;
+    }
+    if (config != null) {
+      return config.getCompatibilityLevel();
+    } else if (returnTopLevelIfNotFound) {
+      config = (ConfigValue) get((K) new ConfigKey(null));
+      return config != null ? config.getCompatibilityLevel() : defaultForTopLevel;
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public Mode mode(String subject,
+                   boolean returnTopLevelIfNotFound,
+                   Mode defaultForTopLevel
+  ) {
+    ModeKey modeKey = new ModeKey(subject);
+    ModeValue modeValue = (ModeValue) get((K) modeKey);
+    if (modeValue == null && subject == null) {
+      return defaultForTopLevel;
+    }
+    if (modeValue != null) {
+      return modeValue.getMode();
+    } else if (returnTopLevelIfNotFound) {
+      modeValue = (ModeValue) get((K) new ModeKey(null));
+      return modeValue != null ? modeValue.getMode() : defaultForTopLevel;
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public boolean hasSubjects(String subject) throws StoreException {
+    return hasSubjects(matchingPredicate(subject));
+  }
+
+  public boolean hasSubjects(Predicate<String> match) throws StoreException {
+    return store.entrySet().stream()
+        .anyMatch(e -> {
+          K k = e.getKey();
+          V v = e.getValue();
+          if (k instanceof SchemaKey) {
+            SchemaKey key = (SchemaKey) k;
+            SchemaValue value = (SchemaValue) v;
+            if (value != null && !value.isDeleted()) {
+              return match.test(key.getSubject());
+            }
+          }
+          return false;
+        });
+  }
+
+  @Override
+  public void clearSubjects(String subject) throws StoreException {
+    clearSubjects(matchingPredicate(subject));
+  }
+
+  public void clearSubjects(Predicate<String> match) throws StoreException {
+    store.entrySet().removeIf(e -> {
```

Comment:
This implementation assumes that the event is triggered only by SR and removes everything from the cache. What happens when someone, intentionally/un-intentionally publishes this event when the Subject is not IMPORT ready?
If the general idea is to clean up the deletes from the cache just so that import works without too much fuzziness in the logic, would it be safer to just act on the deleted schema versions and also rename the event to call it out specifically?

Reply:
Great point!  I'll modify the code to only clear out deleted schemas.